gollm/vertexai.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gollm
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"cloud.google.com/go/vertexai/genai"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/option"
+	"k8s.io/klog/v2"
+)
+
+const (
+	vertexaiDefaultModel = "gemini-2.0-pro-exp-02-05"
+)
+
+// NewVertexAIClient builds a client for the VertexAI API.
+func NewVertexAIClient(ctx context.Context) (*VertexAIClient, error) {
+	log := klog.FromContext(ctx)
+
+	var opts []option.ClientOption
+
+	creds, err := google.FindDefaultCredentials(ctx, "https://www.googleapis.com/auth/generative-language", "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		return nil, fmt.Errorf("finding default credentials: %w", err)
+	}
+	opts = append(opts, option.WithCredentials(creds))
+
+	projectID := ""
+	location := "us-central1"
+
+	if projectID == "" {
+		cmd := exec.CommandContext(ctx, "gcloud", "config", "get", "project")
```

Comment:
Curious about your thoughts on taking a dependency on `gcloud` (even an optional one).

If the design of the plug in to interface with multiple models via their API's it might be best to keep the interfaces as a web service call

Reply:
So the "current project" is only a gcloud concept, there's no GCP API call for this.